sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala

```diff
@@ -62,11 +62,22 @@ trait ConstraintHelper {
    */
   def inferAdditionalConstraints(constraints: Set[Expression]): Set[Expression] = {
     var inferredConstraints = Set.empty[Expression]
-    constraints.foreach {
+    val binaryComparisons = constraints.filter(_.isInstanceOf[BinaryComparison])
+    binaryComparisons.foreach {
       case eq @ EqualTo(l: Attribute, r: Attribute) =>
-        val candidateConstraints = constraints - eq
+        val candidateConstraints = binaryComparisons - eq
         inferredConstraints ++= replaceConstraints(candidateConstraints, l, r)
         inferredConstraints ++= replaceConstraints(candidateConstraints, r, l)
+      case eq @ EqualTo(l @ Cast(lc: Attribute, _, tz), r: Attribute) =>
+        val candidateConstraints = binaryComparisons - eq
+        val bridge = Cast(r, lc.dataType, tz)
+        inferredConstraints ++= replaceConstraints(candidateConstraints, r, l)
+        inferredConstraints ++= replaceConstraints(candidateConstraints, lc, bridge)
```

Comment:
is this safe? cast is not reversible. `cast(int_col as long) = long_col` can have different result than `int_col = cast(long_col as int)`.

Reply:
https://github.com/apache/spark/pull/27252/commits/2060190d319c9008371ae9af8aabb2ebc757dde6: Makes it only support filter at higher data type.